ICSharpCode.Decompiler/IL/ControlFlow/ConditionDetection.cs

```diff
@@ -47,264 +59,546 @@ public class ConditionDetection : IBlockTransform
+		{
+			if (inst is Block block && block.Instructions.Count > 0)
+				inst = block.Instructions.Last();
+
+			if (inst.HasFlag(InstructionFlags.EndPointUnreachable)) {
+				exitInst = inst;
+				return true;
 			}
+			
+			exitInst = null;
 			return false;
 		}
 
 		/// <summary>
-		/// Gets whether <c>potentialBranchInstruction</c> is a branch to a block
-		/// that is dominated by <c>cfgNode</c>.
-		/// If this function returns true, we replace the branch instruction with the block itself.
+		/// Gets the final instruction from a block (or a single instruction) assuming that all blocks
+		/// or instructions in this position have unreachable endpoints
 		/// </summary>
-		bool IsUsableBranchToChild(ControlFlowNode cfgNode, ILInstruction potentialBranchInstruction)
+		private static ILInstruction GetExit(ILInstruction inst)
 		{
-			Branch br = potentialBranchInstruction as Branch;
-			if (br == null)
-				return false;
-			var targetBlock = br.TargetBlock;
-			return targetBlock.Parent == currentContainer
-				&& targetBlock.IncomingEdgeCount == 1 && targetBlock.FinalInstruction.OpCode == OpCode.Nop
-				&& cfgNode.Dominates(context.ControlFlowGraph.GetNode(targetBlock));
+			ILInstruction exitInst = inst is Block block ? block.Instructions.Last() : inst;
+			// Last instruction is one with unreachable endpoint
+			// (guaranteed by combination of BlockContainer and Block invariants)
+			Debug.Assert(exitInst.HasFlag(InstructionFlags.EndPointUnreachable));
+			return exitInst;
 		}
-		
-		private bool IsBranchOrLeave(ILInstruction inst)
+
+		/// <summary>
+		/// Returns true if inst is Nop or a Block with no instructions.
+		/// </summary>
+		private static bool IsEmpty(ILInstruction inst) => 
+			inst is Nop || inst is Block block && block.Instructions.Count == 0 && block.FinalInstruction is Nop;
+
+		/// <summary>
+		/// Import some pattern matching from HighLevelLoopTransform to guess the continue block of loop containers.
+		/// Used to identify branches targetting this block as continue statements, for ordering priority.
+		/// </summary>
+		/// <returns></returns>
+		private Block GuessContinueBlock()
 		{
-			switch (inst) {
-				case Branch branch:
-					return true;
-				case Leave leave:
-					// only void returns are supported as 'exit points'
-					return leave.Value.MatchNop();
-				default:
-					return false;
+			if (currentContainer.Kind != ContainerKind.Loop)
+				return null;
+
+			// continue blocks have exactly 2 incoming edges
+			if (currentContainer.EntryPoint.IncomingEdgeCount == 2) {
+				try {
+					var forIncrement = HighLevelLoopTransform.GetIncrementBlock(currentContainer, currentContainer.EntryPoint);
+					if (forIncrement != null)
+						return forIncrement;
+				} catch (InvalidOperationException) {
+					// multiple potential increment blocks. Can get this because we don't check that the while loop
+					// has a condition first, as we don't need to do too much of HighLevelLoopTransform's job.
```

Comment:
I originally had it return null, but the code looked ugly in comparison, and wasn't necessary inside HighLevelLoopTransform because of preconditions. Avoiding exceptions as part of normal control flow is a design principle I'm aware of, so I'll favour that.